File: frontend/js/vendas.js
const token = localStorage.getItem("token");

if (!token) {
    mostrarMensagem("Faça login para acessar", "erro");
    window.location.href = "login.html";
}

function ir(pagina) {
    window.location.href = pagina;
}

function logout() {
    localStorage.removeItem("token");
    window.location.href = "login.html";
}

let vendaAtual = null;
let itensVenda = [];

async function carregarClientes() {

    const data = await api("/clientes")

    const select = document.getElementById("clientes");

    data.forEach(c => {
        select.innerHTML += `<option value="${c.id}">${c.nome}</option>`;
    });
}

async function carregarProdutos() {

    const data = await api("/produtos/");

    const select = document.getElementById("produtos");

    data.forEach(p => {
        select.innerHTML += `<option value="${p.id}">${p.nome} - R$ ${p.valor}</option>`;
    });
}

function adcionarItem() {

    const selectProduto = document.getElementById("produtos");
    const produto_id = parseInt(selectProduto.value);
    const produto_nome = selectProduto.options[selectProduto.selectedIndex].text;
    const quantidade = parseInt(document.getElementById("quantidade").value);

    if (isNaN(quantidade) || quantidade <= 0) {
        mostrarMensagem("Quantidade inválida", "erro");
        return;
    }

    itensVenda.push({
        produto_id,
        produto_nome,
        quantidade
    });

    renderizarItens();

    document.getElementById("quantidade").value = "";
}

function renderizarItens() {
    const lista = document.getElementById("listaItens");
    lista.innerHTML = "";

    itensVenda.forEach((item, index) => {

        const li = document.createElement("li");

        li.innerHTML = `${item.produto_nome} - Quantidade: ${item.quantidade}
                <button onclick="removerItem(${index})">Remover</button>`;

        lista.appendChild(li);
    });
}

function removerItem(index) {
    itensVenda.splice(index, 1);
    renderizarItens();
}

async function finalizarVenda() {

    const cliente_id = document.getElementById("clientes").value;

    if (itensVenda.length === 0) {
        mostrarMensagem("Adicione pelo menos um item", "erro");
        return;
    }

    try {

        await api("/vendas/", {
            method: "POST",
            body: JSON.stringify({
                cliente_id: parseInt(cliente_id),
                itens: itensVenda.map(item => ({
                    produto_id: item.produto_id,
                    quantidade: item.quantidade
                }))
            })
        });

        mostrarMensagem("Venda criada com sucesso!", "sucesso")

        itensVenda = [];
        renderizarItens();
        listarVendas();

    } catch (error) {

        mostrarMensagem("Erro ao criar venda", "erro");
    }

}

async function listarVendas() {

    const data = await api("/vendas/")

    const tabela = document.getElementById("tabelaVendas");
    tabela.innerHTML = "";

    data.forEach(v => {
        const dataFormatada = new Date(v.data_venda).toLocaleString();

        tabela.innerHTML += `
                    <tr>
                        <td>${v.id}</td>
                        <td>${v.cliente_nome}</td>
                        <td>${v.total}</td>
                        <td>${dataFormatada}</td>
                        <td><span class="status ${v.status}">${v.status}</span></td>
                        <td><button onclick="verDetalhes(${v.id})">Detalhes</button>
                            <button class="btn-danger" onclick="excluirVenda(${v.id})">Excluir</button></td>
                    </tr>`;
    });
}

async function verDetalhes(id) {

    console.log("clicou detalhes", id);

    const data = await api(`/vendas/${id}`)

    vendaAtual = id;

    let html = `
                Cliente: ${data.venda.cliente_nome}<br>
                Total: R$ ${data.venda.total}<br>
                Pago: R$ ${data.total_pago}<br>
                Status: ${data.status}<br><br>
                <b>Itens:</b><br>
                `;

    data.itens.forEach(i => {
        html += `${i.produto_nome} - ${i.quantidade}x (R$ ${i.valor_unitario})
                
                <button onclick="editarItem(${i.produto_id}, ${i.quantidade})">Editar</button>
                <br>`;
    });

    document.getElementById("detalhes").innerHTML = html;
    document.getElementById("modal").style.display = "flex";
}

function fecharModal() {
    document.getElementById("modal").style.display = "none";
}

async function pagar() {

    const valor = document.getElementById("valorPagamento").value;

    try {

        await api(`/vendas/${vendaAtual}/pagamentos`, {
            method: "POST",
            body: JSON.stringify({ valor_pago: parseFloat(valor) })
        });

        mostrarMensagem("Pagamento registrado com sucesso!", "sucesso");

        fecharModal();
        listarVendas();

    } catch (error) {
        mostrarMensagem("Erro ao processar pagamento", "erro");
    }
}

async function fecharMes() {

    const confirmar = confirm("Deseja fechar o caixa do mês?");

    if (!confirmar) return;

    try {

        await api("/vendas/fechar-mes", {
            method: "POST"
        });

        mostrarMensagem("Mês fechado com sucesso!", "sucesso");

        listarVendas();

    } catch (error) {

        mostrarMensagem("Erro ao fechar o mês", "erro");
    }
}

async function editarItem(produto_id, quantidadeAtual) {

    const novaQuantidade = prompt("Digite a nova quantidade:", quantidadeAtual);

    if (novaQuantidade === null) {
        return;
    }

    const quantidade = parseInt(novaQuantidade);

    if (isNaN(quantidade) || quantidade <= 0) {
        mostrarMensagem("Quantidade inválida", "erro");
        return;
    }

    try {

        await api(`/vendas/${vendaAtual}
            /itens/${produto_id}?quantidade=
            ${quantidade}`, {
            method: "PUT",
        });

        mostrarMensagem("Item atualizado com sucesso!", "sucesso");

        verDetalhes(vendaAtual);

        listarVendas();

    } catch (error) {
        mostrarMensagem("Erro ao atualizar item", "erro");
    }

}

async function excluirVenda(id) {

    const confirmar = confirm("Tem certeza que deseja excluir esta venda?")
    if (!confirmar) {
        return;
    }

    try {

        await api(`/vendas/${id}`, {
            method: "DELETE",
        });

        mostrarMensagem("Venda excluída com sucesso!", "sucesso");

        listarVendas();

    } catch (error) {
        mostrarMensagem("Erro ao excluir venda", "erro");
    }
}

window.onload = () => {
    carregarClientes();
    carregarProdutos();
    listarVendas();
};

File: tests/helpers/vendas.py
from helpers.auth import headers

from helpers.clientes import criar_cliente

from helpers.produtos import criar_produto

def criar_venda_completa(client, token):

    cliente = criar_cliente(client, token)
    produto = criar_produto(client, token)

    
    response = client.post(
        "/vendas",
        headers=headers(token),
        json={
            "cliente_id": cliente["id"],
            "itens": [
                {
                    "produto_id": produto["id"],
                    "quantidade": 1,
                }
            ]
        }
    )

    return response.json()["dados"]

def criar_venda_simples(client, token):

    cliente = criar_cliente(client,token)

    response = client.post(
        "/vendas/simples",
        headers=headers(token),
        json={
            "cliente_id": cliente["id"],
            "itens": []
        }
    )

    return response.json()["dados"]

def criar_venda_simples_por_cliente_id(client, token, cliente_id):

    response = client.post(
        "/vendas/simples",
        headers=headers(token),
        json={
            "cliente_id": cliente_id,
            "itens": []
        }
    )

    return response.json()["dados"]


def criar_item_e_adicionar(client, token, venda_id, produto_id):
    
    response = client.post(
        f"/vendas/{venda_id}/itens",
        headers=headers(token),
        json={
            "produto_id": produto_id,
            "quantidade": 1,
        }
    )

    return response.json()["dados"]
def criar_item_e_adicionar_quantidade(client, token, venda_id, produto_id, quantidade):
    
    response = client.post(
        f"/vendas/{venda_id}/itens",
        headers=headers(token),
        json={
            "produto_id": produto_id,
            "quantidade": quantidade
        }
    )

    return response.json()["dados"]



def criar_venda_e_item(client, token):

    venda = criar_venda_simples(client, token)

    produto = criar_produto(client, token)

    item = criar_item_e_adicionar(client, token, venda["id"], produto["id"])

    return venda, item

def buscar_venda_por_id(client, token, venda_id):

    response = client.get(
        f"/vendas/{venda_id}",
        headers=headers(token),
    )

    return response.json()["dados"]

def listar_vendas(client, token):

    response = client.get(
        "/vendas",
        headers=headers(token),
    )

    return response.json()["dados"]

def fazer_pagamento_venda(client, token, venda_id, valor):

    response = client.post(
        f"/vendas/{venda_id}/pagamentos",
        headers=headers(token),
        json={
            "valor_pago": valor
        }
    )
    
    return response.json()["dados"]

def atualizar_item_venda(client, token, quantidade):

    venda_criada = criar_venda_simples(client, token)

    produto_criado = criar_produto(client, token)

    produto_id = produto_criado["id"]

    venda_id = venda_criada["id"]

    criar_item_e_adicionar(client, token, venda_id, produto_id)

    venda = buscar_venda_por_id(client, token, venda_id)

    response = client.put(
        f"/vendas/{venda['id']}/itens/{produto_id}",
        headers=headers(token),
        json={
            "quantidade": quantidade
        }
    )

    return response

def atualizar_item(client,token, venda_id, produto_id, quantidade):

    response = client.put(
        f"/vendas/{venda_id}/itens/{produto_id}",
        headers=headers(token),
        json={
            "quantidade": quantidade
        }
    )

    return response

def deletar_venda(client, token, venda_id):

    response = client.delete(
        f"/vendas/{venda_id}",
        headers=headers(token)
    )

    return response

def fechar_venda_mes(client, token):
    
    response = client.post(
        "/vendas/fechar-mes",
        headers=headers(token)
    )

    return response

def listar_vendas_fechadas(client, token, mes=None, ano=None):

    response = client.get(
        f"/vendas/historico?mes={mes}&ano={ano}",
        headers=headers(token)
    )

    return response.json()["dados"]

def criar_venda_simples_erro(client, token, cliente_id):

    response = client.post(
        "/vendas/simples",
        headers=headers(token),
        json={
            "cliente_id": cliente_id,
            "itens": []
        }
    )

    return response

def criar_item_e_adicionar_erro(client, token, venda_id, produto_id):
    
    response = client.post(
        f"/vendas/{venda_id}/itens",
        headers=headers(token),
        json={
            "produto_id": produto_id,
            "quantidade": 1,
        }
    )

    return response